### DISCUSSION TITLE
let's develop in a develop branch and pin master to the current release

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#799] [Ideas]
We've been onboarding folks to our enterprise app that uses oqtane.  We skip a beat in the setup process where when we clone oqtane.framework and then checkout the latest tagged version (in this case 1.0.4).  Then I explain that that master branch is the working branch and there is no develop branch.  This is a little confusing to folks who expect the master branch to be production ready.

I think having a volatile develop branch and keeping master pinned to the last release would be helpful for new oqtane developers.  This would make oqtane act and feel like most of the other repos. reducing an extra step in the process.

I feel that this change would help the project act and feel like other projects.

optional bonus points:  also rename "master" to "main" in the process.  All the cool kids are doing it. :)

**Comment by hishamco** [comment]
I already requested that long time ago

**Reply by hishamco** [reply]
https://github.com/oqtane/oqtane.framework/issues/313

**Reply by markdav-is** [reply]
thanks for the link and for bring this up previously.  I searched for it in discussions but didn't search in the issues. I'll cross my fingers!

**Reply by PizzaConsole** [reply]
I agree that this would make a lot of sense, but ultimately its up to the repo maintainer @sbwalker ...

**Reply by sbwalker** [reply]
Yes I agree that we need to do this. However there is another dilemma to consider now as well. Oqtane V2 will be based on .NET 5. It will he released on Nov 10. So if we create a dev branch now and it contains the .NET 5 bits, does this mean that main moves to .NET 5 on Nov 10? My preference is to keep things as simple as possible as realistically I am not sure if the community has the bandwidth to support dev and main branches of V1 (.NET Core 3.1) and V2 (.NET 5). Thoughts?

**Reply by mikecasas** [reply]
At least a dev/main of 3.1 would be a good start.

**Reply by hishamco** [reply]
@sbwalker IMHO while the .NET 5.0 not shipped yet, I'm still see a value of creating ` dev` branch and tag all the previous release and pin version `1.4.0` into a master for now, so when we release `2.0.0` we will bump the version in the master and keep `dev` for nightly builds

Please let me know if you need a help on this I can take care of this if you don't have time

Thanks

**Reply by kencutt** [reply]
Hi everyone.....I'm new here and to GIT as well.  I agree we need to think about implementing a DEV branch.  When version 2 comes around, my idea would be to fork the project and leave 1.4.0 forked and still pointing at .NET Core 3.1 and the main fork of 2.0 pointing to .NET 5 with the intent be to "retire" version 1.4 and simply leave it available for users that want to still use .NET Core 3.1 and assist them as bandwidth permits to integrate changes from the main fork.  Perhaps that is overkill?

**Reply by PizzaConsole** [reply]
If the real plan is to move to .NET 5 ASAP, then why don't we fork out the .NET Core 3.1 to its own branch and convert the master to .NET 5 and create just a dev branch for .NET 5

**Reply by sbwalker** [reply]
My personal preference is to do a combination of what many of you suggested above. 

1. Oqtane 1.x / .NET Core 3.1 is currently the main branch
2. Introduce a dev branch which will be focused on Oqtane 2.x / .NET 5 
3. Create a 1.0.5 release to include all of the latest ( non .NET 5 ) code changes
4. (Hopefully) on Nov 10 when .NET 5 is released, merge the dev branch into the main branch and release V2.0. This effectively moves all future development to .NET 5
5. Development continues on in the dev branch and gets merged to main for minor releases ( ie. 2.x )
6. Hotfixes can be implemented in main to address critical bugs and released as revision releases ( ie. 2.1.x ) ( and will also need to be implemented in dev )

So essentially this keeps a simple 2 branch approach. But it also means that Oqtane 1.x /  .NET Core 3.1 maintenance effectively ends on Nov 10.

**Reply by hishamco** [reply]
1. You mean `1.0.4` release?

2. LGTM

3. What should be the list for `1.0.5`?

4. LGTM

5. LGTM

6.  Patches need to be implemented in `dev` after that should they will be a part of `master`?

**Reply by sbwalker** [reply]
After more consideration of this topic, I have some additional thoughts...

- the plan is to release Oqtane V2 to coincide with .NET 5. The release date for .NET 5 is Nov 10. That is barely 3 weeks away.
- in order to prepare for the V2 release in the remaining time frame we really need an "all hands on deck" approach to ensure the product is ready. Based on past experience, if you do not force people to use the new bits, most will choose the path of least resistance and will not participate.
- the master branch currently contains a number of improvements to support Static Localization. These are V2 enhancements and are not intended to be released as part of a V1.x release. This makes the effort to create a 1.0.5 release as I suggested above a bit more challenging.
- as we prepare for V2 we need to focus on stability and ensuring all major use cases work as expected, and reduce focus on other types of enhancements/refactoring
- we have limited resources so in the short term it would be best to avoid any additional administrative overhead or burden which may prevent us from preparing for the V2 release. 
 
As a result of these above items, I am now suggesting a slightly different approach for branching:

1. The current master branch will be used for preparing for V2.0 ( .NET 5 changes have already been merged to affect this change )
2.  Once V2.0 is released, introduce a dev branch 
3.  Development will occur in the dev branch and get merged to master for releases

So I am still advocating for a simple 2 branch approach - its just a modification of the timing of when this branching strategy takes effect.

The only outstanding question is if/what should be done for V1.x? I had suggested previously that a 1.0.5 release may be a good idea. In order to create such a release we would need to seed a branch with V1.0.4 and selectively cherry pick any bug fixes into it. This would be a bit of a challenging exercise... so the question is, would it be valuable? Will people who are running on V1.0.4 today want to stay on .NET Core 3.1? For how long? Based on resources, I do not think it is feasible for Oqtane to actively support both V1.x and V2.x. So, as I mentioned previously, I am envisioning a full transition to V2/.NET 5. Which seems to indicate that a V1.0.5 may not be that beneficial? Thoughts?

**Reply by hishamco** [reply]
What if we create a `1.0.4` branch and create a tag for it, so the people who want to continue with .NET Core 3.1 will not mandatory to install .NET 5.0 RC or upgrade VS, but when `2.0.0` is release our main focus will be on latest bits, and patch releases can be created from the last commit of `1.0.4` release and new branch or tag may created too

**Reply by mikecasas** [reply]
My opinion - don't worry about the 1.x branch. Put the focus on .NET 5 and V2.

**Reply by markdav-is** [reply]
@hishamco there is currently a 1.0.4 tag and our team in pinned to that for a while until 1.0.5 or 2 comes and we move to that.

**Reply by PizzaConsole** [reply]
I agree with @mikecasas. Don't worry about 1.x

**Comment by markdav-is** [comment]
it's up to @sbwalker to pick the path that doesn't burden the main contributors.  I'd bite the bullet when you start working on V2 by doing that work in develop.  master can still be getting "patched" on 3.1 if there is parallel work.   We have not been considering a move to Net5 for our product, but when Oqtane rolls, we'll roll with it.

**Comment by mikecasas** [comment]
I haven't used all the features of Github, but can you give permissions for certain branches in the repo?  Meaning allow certain contributors to approve PRs on the dev branch, but only allow @sbwalker to pull into master/main?

This way the dev branch could move faster and have all the CI/CD needed without disturbing the main branch.

**Reply by hishamco** [reply]
I' still see there's a vale to create a team from very active members instead to make @sbwalker in pressure for handle everything, but the decision back to him

**Reply by sbwalker** [reply]
Like many of you, I have a full time job which means that I need to spend time on Oqtane outside of those hours. So I would like to find a way to get more people involved. 

However to be totally honest, at this point the amount of review and feedback required for PRs is still very high - and many PRs are not valid and not accepted. This is mainly because the community is still learning the details of framework and are often not aware of the intended use cases, architecture, etc... which is where additional discussion is often needed and extra guidance needs to be provided. 

I think there are a number of folks who are on track to get approval rights for PRs however I think a bit more time is needed to ensure optimal review outcomes ( ie. to prevent rollbacks ). In the meantime I will commit to reviewing PRs as quickly and efficiently as I can so as not to be a bottleneck.

**Reply by sbwalker** [reply]
I do think CI/CD and unit tests would be very helpful. They definitely cannot replace architectural code reviews but they are able to catch minor mechanical issues. This is an area where we can definitely have more involvement.

**Reply by hishamco** [reply]
> I do think CI/CD and unit tests would be very helpful.

Disagree, there 're many benefits from using them:

- Reduce the time for code review

- Ensure that the new PRs doesn't break the build

- Improve code quality which leads to improve the quality of the software itself

As I see in ASP.NET Core repos and many OSS that I worked with, I suggested the following at least:

- Setup CI/CD for Windows, Linux & Mac OS

- Add unit tests for almost the new features

- Add functional tests and E2E whenever it's required

Believe me over the time every contributor will add a unit tests for the PRs that he send without asking, and as I mentioned many times I can handle many parts and create unit and functional tests as much as I can if you agree in the principle

Thanks

**Reply by mikecasas** [reply]
@hishamco He said CI/CD and unit tests **WOULD BE** beneficial.

**Comment by mikecasas** [comment]
I started #807 to track items relating to building the dev branch and pushing the images to Dockerhub.

@sbwalker I hope I didn't overstep, but I created an 'oqtane' organization on Dockerhub. Nothing is there now.
https://hub.docker.com/orgs/oqtane
